### PR TITLE
add bootstrap option

### PR DESCRIPTION
**Pull request by romulodl**
Added the bootstrap option (issue #52)

After looking the issue and the tips on the pull request 64. I did quickly this peace of code and added this feature:
- if you add a file called bootstrap.php to the test folder, it will be automatically loaded;
- Exceptions will be raised if you pass a invalid, inexistent or not pass any file (only --bootstrap without the /path/to/file.php)

TODO: read the yaml file for a possible bootstrap configuration and load it.

I would like a review from you guys if the way I've done is ok or if I'm breaking any pattern that you could be using.

Thanks
Romulo

### CODE REVIEW

**Comment by romulodl** [comment]
There are bugs on these code, once I have time I will fix them. I just would like a review to know if I'm following the right path

**Comment by romulodl** [comment]
- Added the --bootstrap=path/to/file.php param
- If the bootstrap option is set up on phpspec.yml, try to load it
- if the path is empty, the file doesn't exist or the file isn't a ".php". It will return an InvalidArgumentException

Thanks
Romulo

**Comment by sivad77** [comment]
Any progress with this pull request?

File: src/PHPSpec2/Console/Application.php [hunk 1]
```diff
@@ -188,6 +192,10 @@ public function __construct($version)
         $c->extend('console.commands', function($c) {
             return new Command\DescribeCommand;
         });
+        
+        $c->extend('console.definitions', function($c) {
+            return new InputOption('bootstrap=/path/to/file.php', null, InputOption::VALUE_REQUIRED, 'Run a bootstrap file before start');
```
Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: The construction of the option is wrong. The first argument is the name of the option, it should be something like

```
new InputOption('bootstrap', null, InputOption::VALUE_REQUIRED, 'Run a bootstrap file before start')
```

see http://api.symfony.com/2.2/Symfony/Component/Console/Input/InputOption.html#method___construct

File: src/PHPSpec2/Console/Application.php [hunk 2]
```diff
@@ -200,6 +208,7 @@ public function getContainer()
      */
     public function doRun(InputInterface $input, OutputInterface $output)
     {
+
```
Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: this line should be removed

File: src/PHPSpec2/Console/Application.php [hunk 3]
```diff
@@ -221,13 +230,48 @@ public function doRun(InputInterface $input, OutputInterface $output)
...
             $this->add($command);
         }
+        
+        $parent_definition = parent::getDefinition();
+        foreach ($this->container->get('console.definitions') as $definition) {
+            $parent_definition->addOption($definition);
+        }
+        
+        // bootstrap
+
```
Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: this line should be removed

File: src/PHPSpec2/Console/Application.php [hunk 4]
```diff
@@ -221,13 +230,48 @@ public function doRun(InputInterface $input, OutputInterface $output)
...
         }
+        
+        $parent_definition = parent::getDefinition();
+        foreach ($this->container->get('console.definitions') as $definition) {
+            $parent_definition->addOption($definition);
+        }
+        
+        // bootstrap
+        
+        $has_param = $input->getParameterOption('--bootstrap');
```
Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: Use camel case for your variables. And you should use `$input->getOption('bootstrap')` instead, and `$bootstrap` may be a better name for this variable, don't you think?

**Comment by benja-M-1** [comment]
Actually this PR duplicates this one https://github.com/phpspec/phpspec2/pull/64

File: src/PHPSpec2/Console/Application.php [hunk 5]
```diff
@@ -221,13 +230,48 @@ public function doRun(InputInterface $input, OutputInterface $output)
...
          && !$input->hasParameterOption('-h')
          && !$input->hasParameterOption('--help')) {
-            $input = new ArrayInput(array('command' => 'run'));
+             $input = new ArrayInput(array('command' => 'run'));
         }
-
+        
         parent::doRun($input, $output);
     }
+
```
Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: remove this line

File: src/PHPSpec2/Console/Application.php [hunk 6]
```diff
@@ -221,13 +230,48 @@ public function doRun(InputInterface $input, OutputInterface $output)
...
+        // end bootstap
 
         if (!($name = $this->getCommandName($input))
          && !$input->hasParameterOption('-h')
          && !$input->hasParameterOption('--help')) {
-            $input = new ArrayInput(array('command' => 'run'));
+             $input = new ArrayInput(array('command' => 'run'));
         }
-
+
```
Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: remove white spaces

File: src/PHPSpec2/Console/Application.php [hunk 7]
```diff
@@ -221,13 +230,48 @@ public function doRun(InputInterface $input, OutputInterface $output)
...
+            } 
+            else require $realfile;
+        }
+        // end bootstap
 
         if (!($name = $this->getCommandName($input))
          && !$input->hasParameterOption('-h')
          && !$input->hasParameterOption('--help')) {
-            $input = new ArrayInput(array('command' => 'run'));
+             $input = new ArrayInput(array('command' => 'run'));
```
Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: this space is useless

File: src/PHPSpec2/Console/Application.php [hunk 8]
```diff
@@ -221,13 +230,48 @@ public function doRun(InputInterface $input, OutputInterface $output)
...
+            
+            if(!is_file($realfile)){
+                throw new InvalidArgumentException("The bootstrap file ({$file}) doesn't exist");
+            }  
+            elseif(pathinfo($realfile, PATHINFO_EXTENSION) !== "php"){
+                throw new InvalidArgumentException("The bootstrap file ({$file}) isn't a valid php file");
+            } 
+            else require $realfile;
+        }
+        // end bootstap
```
Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: this comment has not much interset :)

File: src/PHPSpec2/Console/Application.php [hunk 9]
```diff
@@ -221,13 +230,48 @@ public function doRun(InputInterface $input, OutputInterface $output)
...
+        }
+        
+        // bootstrap
+        
+        $has_param = $input->getParameterOption('--bootstrap'); 
+        
+        // incorrect parameter
+        if(is_null($has_param)) {
+            throw new InvalidArgumentException('The --bootstrap option needs a value');
+        }
```
Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: As the option is required this check is useless

File: src/PHPSpec2/Console/Application.php [hunk 10]
```diff
@@ -221,13 +230,48 @@ public function doRun(InputInterface $input, OutputInterface $output)
...
+        
+        if(!$has_param && $this->container->has("bootstrap")){
+            $file = $this->container->get("bootstrap");
+        }
+        else{
+            $file = $has_param;
+        }
```
Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: If you want to add a default value from the container, then the option should not be required but optional. You should use something like this instead:

``` php
$bootstrap = $input->getOption('bootstrap');

if (null == $bootstrap) {
    $bootstrap = $this->container->get("bootstrap");
}

if (!is_file($bootstrap)) {
    throw new InvalidArgumentException("The bootstrap file ({$bootstrap}) doesn't exist");
}  else if (pathinfo($bootstrap, PATHINFO_EXTENSION) !== "php") {
    throw new InvalidArgumentException("The bootstrap file ({$bootstrap}) isn't a valid php file");
}

require $bootstrap;
```

Comment (kipit) [inline src/PHPSpec2/Console/Application.php]: Strict equality isn’t prefered to compare against null ?

``` php
null === $bootstrap
```

Comment (benja-M-1) [inline src/PHPSpec2/Console/Application.php]: yes it is :)

**Comment by romulodl** [comment]
Looking at the code. I think it is better that the `bootstrap` option should be added to the `run` command, not directly into the `Application` class. But then I don't know what should be the behavior when someone runs `./bin/phpspec --bootstrap=file.php` and the Application assumes that the `run` command was called.

**Comment by romulodl** [comment]
- Add the `bootstrap` to the Run Command.
- The bootstrap file will only be called when you run `./bin/phpspec run --bootstrap="file.php"` and not will be called with `./bin/phpspec --bootstrap="file.php"`

**Pull request closed without merging**